confero-front/src/components/connections/Graph.tsx

```diff
@@ -0,0 +1,177 @@
+import React from "react";
+import { orderBy } from "lodash";
+import {
+  runSimulation,
+  CandidateNode,
+  ConnectionLink,
+  SIM_WIDTH,
+  SIM_HEIGHT,
+  NODE_SIZE
+} from "lib/simulation";
+import { Candidate } from "lib/candidate";
+import { Connection } from "lib/connection";
+import { lookupPartyColor } from "lib/lookups";
+
+interface GraphPropTypes {
+  candidates: Candidate[];
+  connections: Connection[];
+  selectedCandidate: Candidate | null;
+  onSelect: (candidate: Candidate | null) => void;
+}
+
+const Graph: React.FunctionComponent<GraphPropTypes> = ({
+  candidates,
+  connections,
+  selectedCandidate,
+  onSelect
+}) => {
+  const selectedCandidateId = selectedCandidate ? selectedCandidate.id : null;
+  const [simCandidates, setSimCandidates] = React.useState<
+    CandidateNode[] | null
+  >(null);
+
+  const [simEdges, setSimEdges] = React.useState<ConnectionLink[] | null>(null);
+
+  React.useEffect(() => {
+    const { nodes, links } = runSimulation(candidates, connections);
+
+    // Render the graph
+    setSimCandidates(nodes);
+    setSimEdges(links);
+
+    return () => {};
+  }, [candidates, connections]);
+
+  const simCandidatesWithConnections = React.useMemo(() => {
+    if (!selectedCandidateId || !simEdges || !simCandidates)
+      return simCandidates;
+
+    const connectedCandidates = simEdges
+      .filter(({ source }) => source.id === selectedCandidateId)
+      .map(({ target }) => target.id);
+
+    const candidates = simCandidates.map(candidate =>
+      connectedCandidates.includes(candidate.id)
+        ? { ...candidate, connected: true }
+        : { ...candidate, connected: false }
+    );
+
+    // Order: selected candidate, then connected, the rest.
+    return orderBy(candidates, candidate =>
+      candidate.id === selectedCandidateId ? 2 : candidate.connected ? 1 : 0
+    );
+  }, [selectedCandidateId, simEdges, simCandidates]);
+
+  if (!(simCandidatesWithConnections && simEdges)) return null;
+
+  return (
+    <svg
+      style={{ flexGrow: 1, padding: "1rem" }}
+      width="100%"
+      height="100%"
+      viewBox={`0 0 ${SIM_WIDTH} ${SIM_HEIGHT}`}
+      preserveAspectRatio="xMidYMid meet"
+    >
+      {/* Edges */}
+      <g stroke="#999" strokeOpacity={0.25}>
+        {simEdges.map(edge => {
+          if (!renderLink(edge, selectedCandidateId)) return null;
+
+          return (
+            <line
```

Comment:
this was fun - instead of using D3's DOM manipulation to create lines, I'm using D3 to run the simulation, but React to actually render the nodes and lines. I found this a lot more readable, though there was a performance cost